### DISCUSSION TITLE
How should `ForecastingGridSearchCV` behave with `SlidingWindowSplitter` cv strategy?

### DISCUSSION THREAD

**Discussion by acartro** [alan-turing-institute/sktime#3220] [Q&A]
Why when I perform a ForecastingGridSearchCV with a SlidingWindowSplitter validation strategy, the model returned is the model trained with all the train data and not with the last data corresponding to the windows_lenght? 
As I understand it, if I am performing a rolling windows strategy and I am looking for the best model taking into account the windows_lenght, the model that should be returned is the model with the best parameters trained with the latest train values corresponding to the windows_lenght. 
![image](https://user-images.githubusercontent.com/11137446/183636063-78093671-9a37-4ec9-80a8-c8c36f781f72.png)
If I do an evaluate it does return the predictions taking into account the windows_length constraint, but with ForecastingGridSearchCV it does not.

Why is this? I'm not interpreting the idea of rolling windows correctly? doesn't it make more sense to return the model trained with the last windows_lenght and not with all the data?

**Comment by aiwalter** [comment]
> If I do an evaluate it does return the predictions taking into account the windows_length constraint, but with ForecastingGridSearchCV it does not.

this is bcz `evaluate` is used in the grid searches internally for cross-validation. But the grid searches always fit on all data independent to which splitter you are using. I would expect that `sklearn` has same behavior when you run a grid search, the final model will also be fitted on all data and not on a split.

**Reply by acartro** [reply]
> I would expect that sklearn has same behavior when you run a grid search, the final model will also be fitted on all data and not on a split.

And what is the point of doing a rolling windows strategy and then fitting the model with all the data? 
If it represents that we are looking for the best hyperparameters with a windows_length constraint, we will find the best hyperparameters respective to the rolling windows constraint. Therefore, to my understanding, it should return as a model, the model trained with the windows_lenght constraint.

**Reply by aiwalter** [reply]
you are mixing up two things. One thing is how you do cross-validation, the other thing is how you forecast. For cross-validation, there are 3 approaches in forecasting context, `SlidingWindowSplitter`, `ExpandingWindowSplitter` and `SingleWindowSplitter`. All three are implemented in `sktime`. If you want to forecast with a recursive strategy, you can use `make_reduction`.

**Reply by acartro** [reply]
Thank you

**Comment by fkiraly** [comment]
@acartro, @aiwalter, there is a parameter `refit`, by which you can choose what you want to do!

`refit=True` (default) leaves the model in the state where it is fitted on the entire training data. Top half of your picture.

`refit=False` leaves the model in the state where it is fitted on the last `windows_length` window. Bottom half of your picture.

Both strategies are common and sensible (refitting is the default for sklearn grid search, so our default for consistency, too), and we leave the user a choice.

**Reply by fkiraly** [reply]
(the docstring was missing the param in the last stable release, but it actually was there since ages)

**Reply by acartro** [reply]
Thank you, I had not seen this parameter. 

What I see, that with the parameter refit=False from version 0.13.0, it still doesn't return the trained model directly with the last values of the windows_length. You must then use the best_forecaster_ and retrain the model.
Do you know if it will be automated in any version?

Actually, if you want to do the strategy with refit=False you have to do it or I'm wrong:

```
#### library ----
from sktime.datasets import load_airline
from sktime.forecasting.ets import AutoETS
from sktime.forecasting.model_selection import SlidingWindowSplitter, ForecastingGridSearchCV

#### Load data ----
y = load_airline()

#### Fold Strategy ----
FH = 12
WINDOW_LENGTH = 96
STEP_LENGTH = 12
SP=12
cv = SlidingWindowSplitter(step_length=STEP_LENGTH, window_length=WINDOW_LENGTH, fh=list(
    range(1, FH+1)), start_with_window=True)

#### forecaster ----
forecaster = AutoETS(error="multiplicative", trend='additive',
                     seasonal='multiplicative', n_jobs=-1, sp=SP)

#### data ----
y_train = y[y.index < '1960-01']

#### Grid Search ----
gscv = ForecastingGridSearchCV(
    forecaster, cv=cv, param_grid={'error': ['multiplicative'], 'trend': ['additive'],
                                   'seasonal': ['multiplicative']}, strategy='refit', refit=False
)

gscv.fit(y_train)

#### fit model ----
gscv.best_forecaster_.fit(y_train[-WINDOW_LENGTH:])

#### predict model ----
y_pred = gscv.predict(list(
    range(1, FH+1)))
```

And another thing, what is the benefit of doing a rolling windows strategy and then training with all the train data?

**Reply by aiwalter** [reply]
I think there is a bug bcz it raises `NotFittedError` when doing this:
```python
from sktime.datasets import load_shampoo_sales
from sktime.forecasting.model_selection import (
    ExpandingWindowSplitter,
    ForecastingGridSearchCV,
    ExpandingWindowSplitter)
from sktime.forecasting.naive import NaiveForecaster
y = load_shampoo_sales()
fh = [1,2,3]
cv = ExpandingWindowSplitter(
    start_with_window=True,
    fh=fh)
forecaster = NaiveForecaster()
param_grid = {"strategy" : ["last", "mean", "drift"]}
gscv = ForecastingGridSearchCV(
    forecaster=forecaster,
    param_grid=param_grid,
    cv=cv,
    refit=False)
gscv.fit(y)
gscv.predict(fh=[1,2,3])
```
its trying to predict on `self.best_forecaster_ = self.forecaster.clone().set_params(**self.best_params_)` which was not fitted however. The probelm is that `evaluate()` does not return the last fitted forecaster but it can optionally return the `y_train` for each forecaster, so we can take that data to fit the forecaster `if not refit` :). @acartro do you want to work on this? Otherwise I can make a PR

**Reply by aiwalter** [reply]
I created an issue #3234
